### ISSUE TITLE
iOS simulator(safari): how to get the height of the safari native header?

### ISSUE THREAD

**Issue by DudaGod** [appium/appium#14097]
## The problem

Touch actions do not work with my css selectors - [error in appium.log](https://gist.github.com/DudaGod/4304f0958268c825e06eb8f92527df03). The error says that touch API can only be used in the native context. But the native context knows nothing about css selectors.

So I decided to use touch actions using only x and y coordinates. But the problem is that I should calculate Y coordinate taking into account the height of safari native header. Otherwise my touch actions may not hit on the my element.

Screenshot with the native header about which I'm talking:
<img width="450" alt="safari-native-header-big" src="https://user-images.githubusercontent.com/5900697/77254735-fffc8700-6c73-11ea-85ce-ebfb35f26769.png">

And it can be changed after scroll:
<img width="449" alt="safari-native-header-small" src="https://user-images.githubusercontent.com/5900697/77254749-0ee33980-6c74-11ea-933d-02fec927cb10.png">

I tried to run Appium Desktop in order to find the id of the native header and then get its height, but safari in the native context has only 5 elements each of which match on whole view:
<img width="1185" alt="safari-native-elements" src="https://user-images.githubusercontent.com/5900697/77254876-c4ae8800-6c74-11ea-825f-442ab619b097.png">

Native source code:
```
<?xml version="1.0" encoding="UTF-8"?><AppiumAUT><XCUIElementTypeApplication type="XCUIElementTypeApplication" name="Safari" label="Safari" enabled="true" visible="true" x="0" y="0" width="414" height="896">
  <XCUIElementTypeWindow type="XCUIElementTypeWindow" enabled="true" visible="false" x="0" y="0" width="414" height="896">
    <XCUIElementTypeOther type="XCUIElementTypeOther" enabled="true" visible="false" x="0" y="0" width="414" height="896">
      <XCUIElementTypeOther type="XCUIElementTypeOther" enabled="true" visible="false" x="0" y="0" width="414" height="896"/>
    </XCUIElementTypeOther>
  </XCUIElementTypeWindow>
</XCUIElementTypeApplication></AppiumAUT>
```

So how can I get the height of safari native header? Or do you have any ideas how can I use touch actions correctly in web context?

## Environment

* Appium version (or git revision) that exhibits the issue: 1.16.0
* Desktop OS/version used to run Appium: MacOS
* Node.js version (unless using Appium.app|exe): 8.6.0
* Npm or Yarn package manager: npm@6.2.0
* Mobile platform/version under test: Safari 13
* Real device or emulator/simulator: iOS simulator
* Appium CLI or Appium.app|exe: cli

## Link to Appium logs

https://gist.github.com/DudaGod/c6b308940e7078d0174979eabf48abfc

## Code To Reproduce Issue [ Good To Have ]

```
'use strict';

const webdriverio = require('webdriverio');

const browser = webdriverio.remote({
    host: 'localhost',
    port: '4444',
    path: '/wd/hub',
    desiredCapabilities: {
        platformName: 'iOS',
        platformVersion: '13.3',
        version: '13.0',
        deviceName: 'iPhone 11',
        browserName: 'safari',
        automationName: 'XCUITest'
    }
});

(async () => {
    const selector = '.entity-search .carousel .showcase .scroller__wrap';

    try {
        await browser.init();
        await browser.url('http://yandex.ru/search/touch?text=лучшие фильмы 2019');

        const {width} = await browser.getElementSize(selector);
        const {x, y} = await browser.getLocation(selector);

        await browser.touchAction([
            {action: 'press', x: x + width / 2, y: y + 114},
            {action: 'wait', ms: 100},
            {action: 'moveTo', x: -1000, y: y + 114},
            'release'
        ]);

        await browser.pause(2000);

        await browser.end();
    } catch (e) {
        console.log('erorr:', e);
        await browser.end();
    }
})();
```

114 - it's the height of native header which I calculated by the selection method. It will stop working if I set 113.

**Comment by imurchie** [comment]
There should be an element whose name is `'TopBrowserBar'`.
```xml
<XCUIElementTypeOther type="XCUIElementTypeOther" name="TopBrowserBar" enabled="true" visible="true" x="0" y="44" width="414" height="50">
  <XCUIElementTypeOther type="XCUIElementTypeOther" enabled="true" visible="false" x="0" y="0" width="414" height="94">
    <XCUIElementTypeOther type="XCUIElementTypeOther" enabled="true" visible="false" x="0" y="0" width="414" height="94"/>
    <XCUIElementTypeOther type="XCUIElementTypeOther" enabled="true" visible="false" x="0" y="0" width="414" height="94"/>
  </XCUIElementTypeOther>
  <XCUIElementTypeButton type="XCUIElementTypeButton" name="Show Toolbar" label="Show Toolbar" enabled="false" visible="false" x="0" y="0" width="414" height="95"/>
  <XCUIElementTypeOther type="XCUIElementTypeOther" enabled="true" visible="true" x="0" y="44" width="414" height="50">
    <XCUIElementTypeButton type="XCUIElementTypeButton" value="‎localhost" name="URL" label="Address" enabled="true" visible="true" x="19" y="48" width="376" height="36">
      <XCUIElementTypeButton type="XCUIElementTypeButton" name="PageFormatMenuButton" label="Format Options" enabled="true" visible="true" x="9" y="48" width="41" height="36"/>
      <XCUIElementTypeOther type="XCUIElementTypeOther" value="‎localhost" name="URL" label="Address" enabled="true" visible="true" x="172" y="55" width="70" height="21"/>
      <XCUIElementTypeButton type="XCUIElementTypeButton" name="ReloadButton" label="Reload" enabled="true" visible="true" x="368" y="48" width="37" height="36"/>
    </XCUIElementTypeButton>
  </XCUIElementTypeOther>
  <XCUIElementTypeOther type="XCUIElementTypeOther" enabled="true" visible="false" x="0" y="94" width="414" height="1"/>
</XCUIElementTypeOther>
```

**Comment by DudaGod** [comment]
> There should be an element whose name is 'TopBrowserBar'.

I use XCode 11.3.1 with iOS 13.3 and iPhone 11 Pro and I can't see these elements.
Are you get these elements using iOS simulator? And if yes, which versions are you use?

**Comment by imurchie** [comment]
The one posted above is from an iPhone 11 Pro Max simulator, iOS 13.4 with Xcode 11.4b3.

But...

Xcode 11.3 (11C29), iPhone 11 Pro
```xml
<XCUIElementTypeOther type="XCUIElementTypeOther" name="topBrowserBar" enabled="true" visible="true" x="0" y="44" width="375" height="50">
  <XCUIElementTypeOther type="XCUIElementTypeOther" enabled="true" visible="false" x="0" y="0" width="375" height="94">
    <XCUIElementTypeOther type="XCUIElementTypeOther" enabled="true" visible="false" x="0" y="0" width="375" height="94"/>
    <XCUIElementTypeOther type="XCUIElementTypeOther" enabled="true" visible="false" x="0" y="0" width="375" height="94"/>
  </XCUIElementTypeOther>
  <XCUIElementTypeButton type="XCUIElementTypeButton" name="Show Toolbar" label="Show Toolbar" enabled="false" visible="false" x="0" y="0" width="375" height="95"/>
  <XCUIElementTypeOther type="XCUIElementTypeOther" enabled="true" visible="true" x="0" y="44" width="375" height="50">
    <XCUIElementTypeButton type="XCUIElementTypeButton" value="‎localhost" name="URL" label="Address" enabled="true" visible="true" x="19" y="48" width="337" height="36">
      <XCUIElementTypeButton type="XCUIElementTypeButton" name="Format Options" label="Format Options" enabled="true" visible="true" x="9" y="48" width="41" height="36"/>
      <XCUIElementTypeOther type="XCUIElementTypeOther" value="‎localhost" name="URL" label="Address" enabled="true" visible="true" x="152" y="55" width="71" height="21"/>
      <XCUIElementTypeButton type="XCUIElementTypeButton" name="ReloadButton" label="Reload" enabled="true" visible="true" x="329" y="48" width="37" height="36"/>
    </XCUIElementTypeButton>
  </XCUIElementTypeOther>
  <XCUIElementTypeOther type="XCUIElementTypeOther" enabled="true" visible="false" x="0" y="94" width="375" height="1"/>
</XCUIElementTypeOther>
```

Xcode 11.3.1 (11C505), iPhone 11 Pro
```xml
<XCUIElementTypeOther type="XCUIElementTypeOther" name="topBrowserBar" enabled="true" visible="true" x="0" y="44" width="375" height="50">
  <XCUIElementTypeOther type="XCUIElementTypeOther" enabled="true" visible="false" x="0" y="0" width="375" height="94">
    <XCUIElementTypeOther type="XCUIElementTypeOther" enabled="true" visible="false" x="0" y="0" width="375" height="94"/>
    <XCUIElementTypeOther type="XCUIElementTypeOther" enabled="true" visible="false" x="0" y="0" width="375" height="94"/>
  </XCUIElementTypeOther>
  <XCUIElementTypeButton type="XCUIElementTypeButton" name="Show Toolbar" label="Show Toolbar" enabled="false" visible="false" x="0" y="0" width="375" height="95"/>
  <XCUIElementTypeOther type="XCUIElementTypeOther" enabled="true" visible="true" x="0" y="44" width="375" height="50">
    <XCUIElementTypeButton type="XCUIElementTypeButton" value="‎localhost" name="URL" label="Address" enabled="true" visible="true" x="19" y="48" width="337" height="36">
      <XCUIElementTypeButton type="XCUIElementTypeButton" name="Format Options" label="Format Options" enabled="true" visible="true" x="9" y="48" width="41" height="36"/>
      <XCUIElementTypeOther type="XCUIElementTypeOther" value="‎localhost" name="URL" label="Address" enabled="true" visible="true" x="152" y="55" width="71" height="21"/>
      <XCUIElementTypeButton type="XCUIElementTypeButton" name="ReloadButton" label="Reload" enabled="true" visible="true" x="329" y="48" width="37" height="36"/>
    </XCUIElementTypeButton>
  </XCUIElementTypeOther>
  <XCUIElementTypeOther type="XCUIElementTypeOther" enabled="true" visible="false" x="0" y="94" width="375" height="1"/>
</XCUIElementTypeOther>
```

Unfortunately, if Apple is not returning the elements, Appium cannot do anything. Until they are reported on your system an experimentally valid value may be the only approach.

**Comment by DudaGod** [comment]
I wrote code using an element about which you spoke and to my surprise it works. Piece of code:
```
await browser.context('NATIVE_APP');

const selector = `type == 'XCUIElementTypeOther' && name CONTAINS 'topBrowserBar'`;
const {width, height} = await browser.getElementSize(`-ios predicate string:${selector}`);
```

So it means that [wdio method getSource](http://v4.webdriver.io/api/property/getSource.html) works incorrectly, because it returns not all native elements.

After that I tried to open Appium Desktop again and now I saw all native elements, I don’t know why it wasn’t showing up last time:

<img width="1589" alt="safari-native-header-with-appium-desktop" src="https://user-images.githubusercontent.com/5900697/77352788-cbaac880-6d50-11ea-8517-940a9a5c338b.png">

Most likely I did something wrong. In any case, thank you very much. I close it.
